src/handler.rs

```diff
@@ -1,124 +1,330 @@
+use crate::karma::{canonicalize_term, Aliases, Karma, KarmaChange, KarmaGroup, KarmaStyle};
+use crate::parse::line;
 use dazeus::{DaZeusClient, Event, Scope};
-use super::grammar::line;
-use super::karma::{Karma, KarmaStyle, KarmaValue};
-use std::error::Error;
-use rustc_serialize::json::ToJson;
-use std::ascii::AsciiExt;
-
-pub fn handle_karma_events(evt: &Event, dazeus: &DaZeusClient) {
-    match line(&evt[3]) {
-        Ok(changes) => {
-            let totals = get_change_totals(changes);
-            for change in totals {
-                let value = store_karma_change(&change, Scope::network(&evt[0]), dazeus).unwrap();
-                if change.style == KarmaStyle::Notify {
-                    let updown = if change.change.total() < 0 { "decreased" } else { "increased" };
-                    dazeus.reply(&evt, &format!("{} {} the karma of {} to {}", &evt[1], updown, change.term, value.votes.to_string())[..], false);
-                }
-            }
-        }
-        Err(_) => warn!("Got a message I don't understand in '{}/{}' from '{}': {}", &evt[0], &evt[2], &evt[1], &evt[3])
+
+pub fn handle_karma_events(evt: &Event, dazeus: &dyn DaZeusClient) {
+    let parse = line(&evt[3]);
+    match parse {
+        Ok((_, changes)) => register_votes(evt, dazeus, &changes),
+        Err(err) => warn!(
+            "parsing message failed in '{}/{}' from '{}': {} ({})",
+            &evt[0], &evt[2], &evt[1], &evt[3], err
+        ),
     }
 }
 
-pub fn reply_to_karma_command(evt: &Event, dazeus: &DaZeusClient) {
-    let term = &evt[4].trim();
-    if term != &"" {
-        let karma = match KarmaValue::from_dazeus(dazeus, Scope::network(&evt[0]), term) {
-            Ok(karma) => karma,
-            _ => KarmaValue::new(term),
+fn register_votes(evt: &Event, dazeus: &dyn DaZeusClient, karma_changes: &[KarmaChange]) {
+    let scope = Scope::network(&evt[0]);
+    let mut karma_groups = Vec::new();
+    for karma_change in karma_changes {
+        debug!(
+            "registering karma change ({:+}) from '{}' in {}/{} for {:?}",
+            karma_change.votes.total(),
+            &evt[1],
+            &evt[0],
+            &evt[2],
+            karma_change.term
+        );
+        let mut karma = Karma::get_from_dazeus_or_new(dazeus, scope.clone(), &karma_change.term);
+        karma.vote(&karma_change.votes);
+        if let Err(err) = karma.save(scope.clone(), dazeus) {
+            let msg = format!("failed to save karma '{}': {}", karma.term, err);
+            error!("{}", msg);
+            dazeus.reply(&evt, &msg, false);
+            continue;
+        }
+
+        let increase = karma_change.votes.total();
+        let group_idx = karma_groups
+            .iter()
+            .position(|(kg, _, _): &(KarmaGroup, _, _)| kg.karmas.contains_key(&karma.term));
+        if let Some(group_idx) = group_idx {
+            karma_groups[group_idx].1 += increase;
+            karma_groups[group_idx].2 =
```

Comment:
It's very unclear to me what tuple this is. Would a struct with named fields not be more clear?

Reply:
Ack. Will fix.